lib/spack/spack/spec.py

```diff
@@ -2106,11 +2170,71 @@ def from_dict(data):
                 continue
 
             yaml_deps = node[name]['dependencies']
-            for dname, dhash, dtypes in Spec.read_yaml_dep_specs(yaml_deps):
+            for dname, dhash, dtypes, _ in Spec.read_yaml_dep_specs(yaml_deps):
                 deps[name]._add_dependency(deps[dname], dtypes)
 
         return spec
 
+    @staticmethod
+    def from_dict(data):
```

Comment:
Here, the spliced DAG loses its root node if its `dag_hash` matches the `dag_hash` of its `build_spec`.
Using the `dag_hash` is the default when determining which hash to use from the dependencies.
Will the `full_hash` always be an alternative? Will we possibly need to use both hashes in order to preserve build provenance on a round trip to JSON?